### PR TITLE
feat(edge-compute): add stack support

### LINKED ISSUE #115: [edge compute] - Update the agent to support Edge stacks
The Edge agent must be able to retrieve Edge stack related information as part of the periodic short poll request.

The agent must be able to store information related to Edge stacks (identifier, version number) in memory.

It must also be able to retrieve a stack file associated to an Edge stack and apply it.

If a stack is referenced in the memory of the agent but not exposed via short the short poll request anymore, the stack must be deleted.

### PR DESCRIPTION
**Pull request by chiptus**
closes #115

### CODE REVIEW

**Comment by chiptus** [comment]
I'm not sure about the location of the EdgeStackManager and all of its types (maybe it needs to be under docker or edge instead of HTTP?)

File: http/portainerclient/portainer_client.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
+package portainerclient
```
Comment (deviantony) [inline http/portainerclient/portainer_client.go:1]: This file should be relocated inside the `client` package.

**Comment by chiptus** [comment]
few things to think about:
1. dependency of tunnelOperator on edgeStackManager. I would prefer to remove it and then the edgeManager can include the functionality of the stackManager. Maybe we should do one of these:
   - separate polling status from polling for stacks status, then we can do polling for stacks in the edgeManager
  - move polling to the edge manager.
  - pass a callback to tunnelOperator that runs when stacks are found
2. in order to set key from the edgeUIServer we pass the manager itself to the server, which feels a bit circular too. Maybe a callback as well?

File: internal/edge/edge.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+		<-timer1.C
+
+		if !manager.IsKeySet() {
+			log.Printf("[INFO] [main,edge,http] [message: Shutting down Edge UI server as no key was specified after %d minutes]", agent.DefaultEdgeSecurityShutdown)
+			edgeServer.Shutdown()
+		}
+	}()
+}
+
+func (manager *EdgeManager) startRuntimeConfigCheckProcess() error {
```
Comment (deviantony) [inline internal/edge/edge.go:121]:
```suggestion
func (manager *EdgeManager) startEdgeBackgroundProcess() error {
```

File: internal/edge/key.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+	return edgeKey, nil
+}
+
+func encodeKey(edgeKey *edgeKey) string {
+	keyInfo := fmt.Sprintf("%s|%s|%s|%s", edgeKey.PortainerInstanceURL, edgeKey.TunnelServerAddr, edgeKey.TunnelServerFingerprint, edgeKey.EndpointID)
+	encodedKey := base64.RawStdEncoding.EncodeToString([]byte(keyInfo))
+	return encodedKey
+}
+
+func (manager *EdgeManager) retrieveEdgeKey(edgeKey string) (string, error) {
```
Comment (deviantony) [inline internal/edge/key.go:98]: I think that this is not the responsability of the EdgeManager to retrieve the Edge key, this should be kept in the `main.go`.

File: http/tunnel/tunnel.go [hunk 4]
```diff
@@ -122,13 +84,20 @@ func (operator *Operator) ResetActivityTimer() {
 // if needed as well as manage schedules.
 // The second loop will check for the last activity of the reverse tunnel and close the tunnel if it exceeds the tunnel
 // inactivity duration.
-func (operator *Operator) Start() error {
-	if operator.key == nil {
-		return errors.New("missing Edge key")
+func (operator *Operator) Start(portainerURL, endpointID, tunnelServerAddr, tunnelServerFingerprint string) error {
```
[the remark this answers is not in the document]
Comment (chiptus) [reply http/tunnel/tunnel.go:87]: so the tunnel operator (poll service) should be created when the key is set

File: internal/edge/stack.go [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+		errorMessage = err.Error()
+	} else {
+		log.Printf("[DEBUG] [internal,edge,stack] [stack_identifier: %d] [message: stack deployed]", stack.ID, stack.Version)
+	}
+
+	manager.stacks[stack.ID] = stack
+
+	err = manager.httpClient.SetEdgeStackStatus(int(stack.ID), responseStatus, errorMessage)
+	if err != nil {
+		log.Printf("[ERROR] [internal,edge,stack] [message: failed setting edge stack status] [error: %v]", err)
```
Comment (deviantony) [inline internal/edge/stack.go:223]:
```suggestion
		log.Printf("[ERROR] [internal,edge,stack] [message: unable to update Edge stack status] [error: %s]", err)
```

File: internal/edge/stack.go [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+	}
+}
+
+func (manager *StacksManager) deleteStack(stack *edgeStack, stackName, stackFileLocation string) {
+	log.Printf("[DEBUG] [internal,edge,stack] [message: removing stack %d]", stack.ID)
```
Comment (deviantony) [inline internal/edge/stack.go:228]:
```suggestion
	log.Printf("[DEBUG] [internal,edge,stack] [stack_identifier: %d] [message: removing stack]", stack.ID)
```

File: internal/edge/stack.go [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+	err := filesystem.RemoveFile(stackFileLocation)
+	if err != nil {
+		log.Printf("[ERROR] [internal,edge,stack] [message: failed deleting edge stack file] [error: %v]", err)
```
Comment (deviantony) [inline internal/edge/stack.go:231]:
```suggestion
		log.Printf("[ERROR] [internal,edge,stack] [message: unable to delete Edge stack file] [error: %s]", err)
```

File: internal/edge/stack.go [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+		return
+	}
+
+	err = manager.dockerStackService.Remove(stackName)
+	if err != nil {
+		log.Printf("[ERROR] [internal,edge,stack] [message: failed removing stack] [error: %v]", err)
```
Comment (deviantony) [inline internal/edge/stack.go:237]:
```suggestion
		log.Printf("[ERROR] [internal,edge,stack] [message: unable to remove stack] [error: %s]", err)
```

File: cmd/agent/main.go [hunk 9]
```diff
@@ -87,41 +86,41 @@ func main() {
 		defer clusterService.Leave()
 	}
 
-	var tunnelOperator agent.TunnelOperator
-	var edgeStackManager agent.EdgeStackManager
+	edgeManager, err := edge.NewManager(options, advertiseAddr, clusterService, infoService)
```
[the remark this answers is not in the document]
Comment (chiptus) [reply cmd/agent/main.go:89]: but then I need to pass them around to anywhere we're initializing the edge manager (HTTP handlers, edge UI server...)

[the remark this answers is not in the document]
Comment (deviantony) [reply cmd/agent/main.go:89]: True, I missed that. Agree with you then.

File: agent.go [hunk 10]
```diff
@@ -232,4 +220,6 @@ const (
 	EdgeKeyFile = "agent_edge_key"
 	// DockerBinaryPath is the path of the docker binary
 	DockerBinaryPath = "/app"
+	// EdgeStackFilesPath is the path where edge stack files are saved
+	EdgeStackFilesPath = "/tmp/edge_stacls"
```
Comment (deviantony) [inline agent.go:224]:
```suggestion
	EdgeStackFilesPath = "/tmp/edge_stacks"
```

**Pull request merged**